addons/product/migrations/13.0.1.2/pre-migration.py

```diff
@@ -123,6 +124,36 @@ def insert_missing_product_template_attribute_value(env):
     )
 
 
+def merge_variants_with_same_attributes(env):
+    openupgrade.logged_query(
+        env.cr, """
+        WITH templates AS (
+            SELECT pt.id, pp.combination_indices, COUNT(*)
+            FROM product_product pp
+            JOIN product_template pt ON pt.id = pp.product_tmpl_id
+            WHERE pp.combination_indices IS NOT NULL
+            GROUP BY pt.id, pp.combination_indices
+            HAVING COUNT(*) > 1
+        )
+        SELECT pp.id, pp.product_tmpl_id
+        FROM product_product pp
+        JOIN product_template pt ON pt.id = pp.product_tmpl_id
+        JOIN templates ON (templates.id = pt.id
+            AND templates.combination_indices = pp.combination_indices)"""
+    )
+    templates = {}
+    for product_id, template_id in env.cr.fetchall():
+        if template_id in templates:
+            templates[template_id] += [product_id]
+        else:
+            templates[template_id] = [product_id]
```

Comment:
I think  it is looks better.
```suggestion
         templates.setdefault(template_id, []).append(product_id)
```